### PR TITLE
digest: introduce FFI bindings for xxHash32/64

### LINKED ISSUE #2003: Embed xxHash  into box.digest
XXH64 and/or XXH32

### PR DESCRIPTION
**Pull request by olegrok**
This patch introduces new hash types for digest module - xxHash32
and xxHash64.

Closes #2003

@TarantoolBot document
Title: digest module supports xxHash32/64

```lua
-- Examples below demonstrate xxHash32.
-- xxHash64 has exactly the same interface

-- Calculate the 32-bits hash (default seed is 0).
digest.xxhash32(string[, seed])

-- Streaming
-- Start a new hash by initializing state with a seed.
-- If no value provided, 0 is used as default.
xxhash = digest.xxhash32.new([seed])
-- Also it's possible to specify seed manually. If no value
-- provided a value initially passed to "new" is used.
-- Here and below "seed" expected to be unsigned
-- number. Function returns nothing.
xxhash:clear([seed])
-- Feed the hash state by calling "update" as many times as
-- necessary. Function returns nothing.
xxhash:update('string')
-- Produce a hash value.
xxhash:result()
```

### CODE REVIEW

File: src/lua/digest.lua [hunk 1]
```diff
@@ -267,4 +311,74 @@ m['aes256cbc'] = {
...
+    local update_fn_name = 'XXH' .. var .. '_update'
+    local digest_fn_name = 'XXH' .. var .. '_digest'
+    local reset_fn_name = 'XXH' .. var .. '_reset'
+    local copy_fn_name = 'XXH' .. var .. '_copyState'
+    local methods = {
+        update = function(self, str)
+            if type(str) ~= 'string' then
+                error("Usage xxhash" .. var .. ":update(string)")
+            end
+            return ffi.C[update_fn_name](self, str, #str) == ffi.C.XXH_OK
```
Comment (Gerold103) [inline src/lua/digest.lua:337]: It makes sense to cache `ffi.C.<name>`. To avoid 2 index operations. Although I see it is not done in the other places, so maybe we could do it in a separate patch later or before this one. Up to you.

File: src/lua/digest.lua [hunk 2]
```diff
@@ -267,4 +311,74 @@ m['aes256cbc'] = {
...
+        end,
+        __tostring = function(self)
+            return string.format('value: %s', self:result())
+        end,
+    })
+
+    xxHash = {
+        new = function()
+            return ffi.new(XXH_state_t)
+        end
```
Comment (Gerold103) [inline src/lua/digest.lua:368]: It looks counterintuitive that I need to call `clear()` right after `new()` just to set the seed. Better add it as an optional parameter for `new`, like pmur does.

File: test/app/digest.result [hunk 3]
```diff
@@ -636,3 +629,173 @@ res
...
 ...
+--
+-- gh-2003 xxHash
+--
+xxhash32 = digest.xxhash32.new()
+---
+...
+xxhash32
+---
+- 'value: 46947589'
```
Comment (Gerold103) [inline test/app/digest.result:640]: Why this value? Isn't it random? Because I don't see where the state is initialized, so the state must be garbage until you call `clear()`. Which also might be an issue. Probably you must call `clear()` internally always inside of `new()`.

**Review by Gerold103 — COMMENTED** [review]
```
    ```lua
    -- Exaples below demonstrate xxHash32.
    -- xxHash64 has exactly the same interface
 ```
"Exaples" -> "Examples".

Why do you need this patch? Is this hash faster or what are the pros?

**Comment by olegrok** [comment]
> Why do you need this patch? Is this hash faster or what are the pros?

Thanks for your review. Seems yes, it's faster. At least according their benchmarks - https://github.com/facebook/zstd/blob/dev/lib/common/xxhash.h#L22-L34

File: src/lua/digest.lua (hunk 1 above)
Comment (olegrok) [reply src/lua/digest.lua:337]: According LuaJIT FFI manual it's wrong approach - https://luajit.org/ext_ffi_tutorial.html . See "To Cache or Not to Cache
" section. It's better to cache "ffi.C" than "ffi.C.<name>".

File: test/app/digest.result (hunk 3 above)
Comment (olegrok) [reply test/app/digest.result:640]: ffi.new fills structure with zero bytes. It seems to be predictable. However, it's reasonable to choose some default seed as e.g. for murmur.
```
tarantool> require('digest').murmur
---
- new: 'function: 0x010ea89a90'
  default_seed: 13
...
```

File: src/lua/digest.lua (hunk 2 above)
Comment (olegrok) [reply src/lua/digest.lua:368]: Agree, I'll fix it

**Comment by olegrok** [comment]
Simple benchmark
```lua
local str = string.rep('x', 10000)

local digest = require('digest')
local clock = require('clock')

local count = 1e5
local start = clock.time()
for _ = 1, count do
    digest.murmur(str)
end
print('digest.murmur(str)', clock.time() - start)

collectgarbage()
collectgarbage()

local start = clock.time()
for _ = 1, count do
    digest.xxhash64(str)
end
print('digest.xxhash64(str)', clock.time() - start)

collectgarbage()
collectgarbage()

local start = clock.time()
local instance = digest.murmur.new()
for _ = 1, count do
    instance:update(str)
end
local _ = instance:result()
print('digest.murmur.new()', clock.time() - start)

local start = clock.time()
local instance = digest.xxhash64.new()
for _ = 1, count do
    instance:update(str)
end
local _ = instance:result()
print('digest.xxhash64.new()', clock.time() - start)

os.exit(0)
--[[
./src/tarantool simple_bench.lua
digest.murmur(str)	1.2278580665588
digest.xxhash64(str)	0.068238019943237
digest.murmur.new()	1.2747230529785
digest.xxhash64.new()	0.063662052154541
--]]
```

File: test/app/digest.result (hunk 3 above)
Comment (olegrok) [reply test/app/digest.result:640]: I made seed as optional value and by default it's 0 (https://github.com/tarantool/tarantool/pull/5828/commits/776c325334459c4b867537a417043062b8a5e874)

File: src/lua/digest.lua [hunk 4]
```diff
@@ -267,4 +311,74 @@ m['aes256cbc'] = {
     end
 }
 
+local function check_unsigned(value)
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/digest.lua]: Let's just remove this check. FFI will return something like "cannot convert 'string' to 'unsigned int'". I hope it's enough.

File: src/lua/digest.lua (hunk 1 above)
Comment (olegrok) [reply src/lua/digest.lua]: Consider https://github.com/tarantool/tarantool/pull/5828/commits/0d546adae6f7727589c67c202e09409666324e9a

File: src/lua/digest.lua (hunk 2 above)
Comment (olegrok) [reply src/lua/digest.lua:363]: Currently seed is an optional value and it's value 0. Description is updated.

File: src/lua/digest.lua [hunk 5]
```diff
@@ -163,7 +209,7 @@ end
 local m = {
     base64_encode = function(bin, options)
         if type(bin) ~= 'string' or
-           options ~= nil and type(options) ~= 'table' then
+                options ~= nil and type(options) ~= 'table' then
```
Comment (Gerold103) [inline src/lua/digest.lua:212]: Unnecessary change, lets keep it as it is.

File: src/lua/digest.lua [hunk 6]
```diff
@@ -267,4 +313,70 @@ m['aes256cbc'] = {
...
+        __tostring = function(self)
+            return string.format('value: %s', self:result())
+        end,
+    })
+
+    xxHash = {
+        new = function(seed)
+            local value = ffi.new(XXH_state_t)
+            value:clear(seed)
+            return value
```
Comment (Gerold103) [inline src/lua/digest.lua:362]: From what I see in pmur code, `default_seed` is stored as an attribute so as to update it in `new(seed)` in case `seed` is specified. Maybe you need to do the same.

File: src/lua/digest.lua [hunk 7]
```diff
@@ -267,4 +313,70 @@ m['aes256cbc'] = {
...
+        end,
+        result = function(self)
+            return builtin[digest_fn_name](self)
+        end,
+        clear = function(self, seed)
+            if seed == nil then
+                seed = self.default_seed
+            end
+            return builtin[reset_fn_name](self, seed) == builtin.XXH_OK
+        end,
```
Comment (Gerold103) [inline src/lua/digest.lua:340]: Would be a good idea to add some empty lines between the function definitions to simplify the reading.

File: src/lua/digest.lua [hunk 8]
```diff
@@ -267,4 +313,70 @@ m['aes256cbc'] = {
...
+            return copy
+        end,
+    }
+
+    local XXH_state_t = ffi.typeof('XXH' .. var .. '_state_t')
+    ffi.metatype(XXH_state_t, {
+        __index = function(_, key)
+            return methods[key]
+        end,
+        __tostring = function(self)
```
Comment (Gerold103) [inline src/lua/digest.lua:353]: Do you really need `__tostring()`? I don't see it used in other hashes.

File: src/lua/digest.lua [hunk 9]
```diff
@@ -267,4 +313,70 @@ m['aes256cbc'] = {
...
+    local update_fn_name = 'XXH' .. var .. '_update'
+    local digest_fn_name = 'XXH' .. var .. '_digest'
+    local reset_fn_name = 'XXH' .. var .. '_reset'
+    local copy_fn_name = 'XXH' .. var .. '_copyState'
+    local methods = {
+        default_seed = 0,
+
+        update = function(self, str)
+            if type(str) ~= 'string' then
+                error("Usage xxhash" .. var .. ":update(string)")
```
Comment (Gerold103) [inline src/lua/digest.lua:328]: I would propose to raise the error in the caller's context, via the second argument of `error()` function. To raise it with the caller's file and line, which should be more useful. Up to you.

**Review by Gerold103 — COMMENTED** [review]
The commit message still is not fixed. `Exaples` -> `Examples`.

File: src/lua/digest.lua (hunk 1 above)
Comment (Gerold103) [reply src/lua/digest.lua]: Nice, didn't know this. Might explain some issues I have in my own patch in another ticket.

File: src/lua/digest.lua (hunk 5 above)
Comment (olegrok) [reply src/lua/digest.lua]: it was a misprint. Fixed.

File: src/lua/digest.lua (hunk 6 above)
Comment (olegrok) [reply src/lua/digest.lua]: Well, I add "default_seed" attribute to `digest.xxhash.new()` table.

File: src/lua/digest.lua (hunk 7 above)
Comment (olegrok) [reply src/lua/digest.lua:340]: Yes, fixed.

File: src/lua/digest.lua (hunk 8 above)
Comment (olegrok) [reply src/lua/digest.lua]: It used to show readable value in console. Removed.

File: src/lua/digest.lua (hunk 9 above)
Comment (olegrok) [reply src/lua/digest.lua]: Changed to error(..., 2).

```
./src/tarantool file.lua 
LuajitError: file.lua:2: Usage digest.xxhash32(string[, unsigned number])
```

File: src/lua/digest.lua (hunk 5 above)
Comment (Gerold103) [reply src/lua/digest.lua]: I still see this change in the first commit.

File: src/lua/digest.lua [hunk 10]
```diff
@@ -267,4 +313,67 @@ m['aes256cbc'] = {
...
+            local copy = xxHash.new()
+            builtin[copy_fn_name](copy.value, self.value)
+            return copy
+        end,
+    }
+
+    local XXH_state_t = ffi.typeof('XXH' .. var .. '_state_t')
+
+    xxHash = {
+        default_seed = 0,
```
Comment (Gerold103) [inline src/lua/digest.lua:352]: You can delete this member and inline `0` in its single usage place.

File: src/lua/digest.lua [hunk 11]
```diff
@@ -267,4 +313,67 @@ m['aes256cbc'] = {
...
+
+        new = function(seed)
+            local self = setmetatable({}, { __index = methods })
```
Comment (Gerold103) [inline src/lua/digest.lua:355]: It makes sense to have a single metatable for all objects. Now you only have single `__index`, but not the entire metatable.

**Review by Gerold103 — COMMENTED** [review]
This is a third time I tell - you still have the typo in the commit message. See the comments above.

File: src/lua/digest.lua [hunk 12]
```diff
@@ -267,4 +313,67 @@ m['aes256cbc'] = {
...
+    local xxHash
+
+    local update_fn_name = 'XXH' .. var .. '_update'
+    local digest_fn_name = 'XXH' .. var .. '_digest'
+    local reset_fn_name = 'XXH' .. var .. '_reset'
+    local copy_fn_name = 'XXH' .. var .. '_copyState'
+    local methods = {
+        update = function(self, str)
+            if type(str) ~= 'string' then
+                error("Usage xxhash" .. var .. ":update(string)", 2)
```
Comment (Gerold103) [inline src/lua/digest.lua:326]: This error is not covered with a test.

File: src/lua/digest.lua (hunk 5 above)
Comment (olegrok) [reply src/lua/digest.lua]: Sorry Vlad, my bad. Fixed. Typo in commit message is also fixed.

File: src/lua/digest.lua (hunk 10 above)
Comment (olegrok) [reply src/lua/digest.lua]: I deleted "default_seed" field from "xxHash" table.

File: src/lua/digest.lua (hunk 12 above)
Comment (olegrok) [reply src/lua/digest.lua]: Fixed

File: src/lua/digest.lua (hunk 11 above)
Comment (olegrok) [reply src/lua/digest.lua]: I plan to do the same as for murmur. But your comment is reasonable. I removed metatable in this place.

File: changelogs/unreleased/add-xxhash-to-digest-module.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+## feature/lua/digest
+
+ * Introduce new hash types in digest module - `xxhash32` and `xxhash64`.
```
Comment (Gerold103) [inline changelogs/unreleased/add-xxhash-to-digest-module.md:3]: You should also add the issue link. See the other changelog entries for examples.

File: src/lua/digest.lua [hunk 14]
```diff
@@ -267,4 +313,70 @@ m['aes256cbc'] = {
...
+    end
+
+    local function copy(self)
+        local copy = xxHash.new()
+        builtin[copy_fn_name](copy.value, self.value)
+        return copy
+    end
+
+    local XXH_state_t = ffi.typeof('XXH' .. var .. '_state_t')
+    local default_seed = 0
```
Comment (Gerold103) [inline src/lua/digest.lua:349]: But it is not much different from what was before. `default_seed` is an upvalue caught in the `new` closures below. You just need to inline it. Otherwise it will be an upvalue regardless of what you do. Although there is a tiny change the Lua interpreter somehow inlines it if detects it is never changed, however I wouldn't rely on it.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the fixes! We are almost there.

```
    -- Streaming
    -- Start a new hash by initializing state with a seed.
    -- If no value provided, 0 is used as default.
    xxhash = digest.xxhash32.new([seed])
    -- Also it's possible to specify seed manually. If no value
    -- provided default 0 is used.
```
It seems you said about seed and its default value 2 times. Although I am not sure which comment to which function refers - the one above it or one below it. Even if it refers to the function below (`clear()`) then it is not true, because when seed is omitted in `clear()`, it uses the seed used previously in `new()`.

```
    -- Function returns "true" in case of success and "false"
```
How can it fail? What are the reasons? And does it set `box.error()`? I know it does not, but this either must be fixed, or reflected in the documentation, because usually all the public functions either throw a Lua `error()`, or use `box.error()`, or set one using `box.error.new()`.

I looked into the sources and it seems it can only fail if the arguments are invalid. For instance, the input string is `NULL`. I don't think it can happen here, we should not bother users with that. The input is guaranteed not to be `NULL`.

File: changelogs/unreleased/add-xxhash-to-digest-module.md (hunk 13 above)
Comment (olegrok) [reply changelogs/unreleased/add-xxhash-to-digest-module.md]: Fixed

File: src/lua/digest.lua (hunk 14 above)
Comment (olegrok) [reply src/lua/digest.lua]: Ok, fixed

**Comment by olegrok** [comment]
> Thanks for the fixes! We are almost there.
> It seems you said about seed and its default value 2 times. Although I am not sure which comment to which function refers - the one above it or one below it. Even if it refers to the function below (clear()) then it is not true, because when seed is omitted in clear(), it uses the seed used previously in new().

Thanks for review. I rephrased commit message a bit.

> I looked into the sources and it seems it can only fail if the arguments are invalid. For instance, the input string is NULL. I don't think it can happen here, we should not bother users with that. The input is guaranteed not to be NULL.

You are right, let's remove this part at all then. Currently update/clear functions return nothing.

**Review by Gerold103 — APPROVED** [review]

File: test/app/digest.test.lua [hunk 15]
```diff
@@ -203,3 +201,66 @@ _ = fiber.create(test_pbkdf2)
...
+
+--
+-- gh-2003 xxHash
+--
+xxhash32 = digest.xxhash32.new()
+xxhash32:result()
+xxhash64 = digest.xxhash64.new()
+xxhash64:result()
+
+-- new takes seed optionally
```
Comment (LeonidVas) [inline test/app/digest.test.lua:213]: From https://github.com/tarantool/tarantool/wiki/Code-review-procedure : "Start sentences from a capital letter, end with a **dot**." Here and below.

File: src/lua/digest.lua [hunk 16]
```diff
@@ -280,4 +324,69 @@ m['aes256cbc'] = {
...
+            }
+            self:clear(self.default_seed)
+            return self
+        end,
+    }
+
+    local call_fn_name = 'XXH' .. var
+    setmetatable(xxHash, {
+        __call = function(_, str, seed)
+            if type(str) ~= 'string'  then
```
Comment (LeonidVas) [inline src/lua/digest.lua:379]: Remove one space before `then`.

**Review by LeonidVas — APPROVED** [review]
Hi! Thank you for the patch.
JFYI from SOP: "
After sending your patch to the mailing list, you should push it on a branch in tarantool github repository. Please give your branch a descriptive name:
<login>/**gh-**XXXX-few-words
"
Generally LGTM, but please answer a few questions.

File: src/lua/digest.lua (hunk 6 above)
Comment (LeonidVas) [reply src/lua/digest.lua]: As far as I understand, "0" in these two places are related values. Maybe you should use an attribute for this?
```
 default_seed = seed or 0,
```
```
 if seed == nil then
    seed = 0
end
```

File: test/app/digest.test.lua (hunk 15 above)
Comment (olegrok) [reply test/app/digest.test.lua]: Fixed. Thanks!

File: src/lua/digest.lua (hunk 16 above)
Comment (olegrok) [reply src/lua/digest.lua]: Fixed. Thanks!

File: test/app/digest.test.lua [hunk 17]
```diff
@@ -203,3 +201,66 @@ _ = fiber.create(test_pbkdf2)
 _ = sentry:get()
 _ = sentry:get()
 res
+
+--
```
[the remark this answers is not in the document]
Comment (olegrok) [reply test/app/digest.test.lua:205]: I don't know but such comments quite often are used among our tests.

File: src/lua/digest.lua [hunk 18]
```diff
@@ -280,4 +324,69 @@ m['aes256cbc'] = {
     end
 }
 
+for _, var in ipairs({'32', '64'}) do
+    local xxHash
+
+    local update_fn_name = 'XXH' .. var .. '_update'
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/digest.lua]: Changed to string.format

File: src/lua/digest.lua [hunk 19]
```diff
@@ -31,6 +31,50 @@ ffi.cdef[[
     void PMurHash32_Process(uint32_t *ph1, uint32_t *pcarry, const void *key, int len);
     uint32_t PMurHash32_Result(uint32_t h1, uint32_t carry, uint32_t total_length);
     uint32_t PMurHash32(uint32_t seed, const void *key, int len);
+
+    /* from third_party/zstd/lib/common/xxhash.c */
+    typedef enum { XXH_OK=0, XXH_ERROR } XXH_errorcode;
```
[the remark this answers is not in the document]
Comment (olegrok) [reply src/lua/digest.lua:36]: Yes, it's a copy-paste. Let's stay as is.

File: test/app/digest.test.lua (hunk 17 above)
[the remark this answers is not in the document]
Comment (LeonidVas) [reply test/app/digest.test.lua:205]: I know (and not only in tests) and I don't object, but I don't understand where it comes from).

File: src/lua/digest.lua (hunk 6 above)
Comment (olegrok) [reply src/lua/digest.lua]: I inlined it as Vlad suggested. See comment - https://github.com/tarantool/tarantool/pull/5828#discussion_r593508162

**Comment by Totktonada** [comment]
It breaks build with `-DENABLE_BUNDLED_ZSTD=OFF`:

```
[ 59%] Linking CXX executable tarantool
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xd80): undefined reference to `XXH32'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xd88): undefined reference to `XXH32_copyState'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xd90): undefined reference to `XXH32_digest'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xd98): undefined reference to `XXH32_reset'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xda0): undefined reference to `XXH32_update'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xda8): undefined reference to `XXH64'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xdb0): undefined reference to `XXH64_copyState'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xdb8): undefined reference to `XXH64_digest'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xdc0): undefined reference to `XXH64_reset'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../x86_64-pc-linux-gnu/bin/ld: CMakeFiles/tarantool.dir/exports.c.o:(.data.rel+0xdc8): undefined reference to `XXH64_update'
collect2: error: ld returned 1 exit status
```

I guess those functions are not public (please, correct me if I'm wrong; I just searched over the [manual](https://facebook.github.io/zstd/zstd_manual.html)). Can we dynamically test those symbols for availability and provide corresponding Lua function only when possible?

**Comment by Totktonada** [comment]
BTW, those symbols are present in `/usr/lib64/libxxhash.so` (dev-libs/xxhash package on Gentoo).
